rotkehlchen/balances/historical.py

```diff
@@ -99,63 +265,230 @@ def get_assets_amounts(
+        event ordering. Each cursor fetches one table, allowing streaming iteration.
+
+        TODO: Refactor this once trades are migrated to history events.
+        """
+        events: list[MinimalHistoryEvent | MinimalTrade] = []
+        with self.db.conn.read_ctx() as cursor1, self.db.conn.read_ctx() as cursor2:
+            main_currency = self.db.get_setting(cursor=cursor1, name='main_currency')
+            trade_filters: list[DBFilter] = [
+                DBNotEqualFilter(and_op=True, column='location', value='B'),
+                DBTimestampFilter(and_op=True, from_ts=from_ts, to_ts=to_ts),
+            ]
+            event_filters: list[DBFilter] = [DBTimestampFilter(
+                and_op=True,
+                to_ts=to_ts,
+                from_ts=from_ts,
+                scaling_factor=FVal(1000),
+            )]
+
+            if assets is not None:
+                asset_identifiers = [asset.identifier for asset in assets]
+                trade_filters.append(DBNestedFilter(
+                    and_op=False,
+                    filters=[
+                        DBMultiStringFilter(and_op=False, column='base_asset', values=asset_identifiers, operator='IN'),  # noqa: E501
+                        DBMultiStringFilter(and_op=False, column='quote_asset', values=asset_identifiers, operator='IN'),  # noqa: E501
+                    ],
+                ))
+                event_filters.append(DBMultiStringFilter(
+                    and_op=True,
+                    column='asset',
+                    values=asset_identifiers,
+                    operator='IN',
+                ))
+
+            trade_where_clauses = []
+            trade_bindings: list[Any] = []
+            for filter_ in trade_filters:
+                clauses, bindings = filter_.prepare()
+                trade_where_clauses.extend(clauses)
+                trade_bindings.extend(bindings)
+
+            trades_iterator = cursor1.execute(
+                f'SELECT timestamp, id, type, base_asset, quote_asset, amount, rate '
+                f'FROM trades WHERE {" AND ".join(trade_where_clauses)} ORDER BY timestamp',
+                trade_bindings,
             )
-            main_currency = self.db.get_setting(cursor=cursor, name='main_currency')
+
+            event_where_clauses = []
+            event_bindings: list[Any] = []
+            for filter_ in event_filters:
+                clauses, bindings = filter_.prepare()
+                event_where_clauses.extend(clauses)
+                event_bindings.extend(bindings)
+
+            events_iterator = cursor2.execute(
+                f'SELECT CAST(timestamp / 1000 AS INTEGER) AS timestamp, event_identifier, asset, amount, type, subtype '  # noqa: E501
+                f'FROM history_events WHERE {" AND ".join(event_where_clauses)} ORDER BY timestamp',  # noqa: E501
+                event_bindings,
+            )
+
+            current_event = next(events_iterator, None)
```

Comment:
So this gets events and trades one by one from the DB? To not hate it all in memory, right? Makes sense ... but would it not be better to have a rolling window of `N` events.

This way you won't have 99,999 DB queries for 99,999 eventsd. Just get them in batches.